manpages/gitcher.1
.\" Manpage for gitcher.
.\" Contact garaje@glezseoane.es to any requirement.
.TH man 1 "18 Apr 2019" "3.2" "gitcher man page"
.SH NAME
gitcher \- the git profile switcher
.SH SYNOPSIS
\fBgitcher\fR [\fB-l\fR|\fB-o\fR|\fB-s\fR|\fB-g\fR|\fB-a\fR|\fB-d\fR ...]
.SH DESCRIPTION
\fBgitcher\fR is a git profile switcher. It facilitates the switching between git profiles, importing configuration settings such as name, email and user signatures.
.SH OPTIONS
.IP "\fINONE\fR"
Opens the interactive mode. It is possible to access to all the operations with this mode, that offers you inline help.
.IP "\fB\-l\fR"
Shows a list with all the \fBgitcher\fR saved profiles.
.IP "\fB\-o\fR"
Displays the activated (ON) gitcher profile for the current working directory.
.IP "\fB\-s\fR \fIprofname\fR"
Set into the existing repository of the current working directory the selected profile.
.IP "\fB\-g\fR \fIprofname\fR"
Set globally the selected gitcher profile.
.IP "\fB\-a\fR \fIprofname\fR \fIname\fR \fIemail\fR \fIsignkey\fR|\fINone\fR \fITrue\fR|\fIFalse\fR
Add a new profile. Inputs are profile name, git user name, git user email, PGP sign key or None (depending if you want to use one), and True or False (depending if you want to use your PGP key to autosign every commit).
.IP "\fB\-d\fR \fIprofname\fR"
Delete the selected profile.
.SH PGP KEYS
\fBgitcher\fR only needs your key ID (the last eight digits of your validation fingerprint) to work. This is the information that you have to provide to \fBgitcher\fR while the creation of your profile.
.SH SAVED DATA
\fBgitcher\fR works with a dotfile saved on user $HOME directory. It is named \fI~/.cherfile\fR and it is not recommended to edit it manually.
.SH EXIT STATUS
Exits 0 on success and 1 on error.
.SH SEE ALSO
\fBgit\fR(1).
.SH BUGS
No known bugs.
.SH AUTHOR
Borja Gonzalez Seoane <garaje@glezseoane.es> is the author of \fBgitcher\fR.
.SH ACKNOWLEDGMENTS
Santiago Fernandez Gonzalez for an effusive reception of the tool, being the highest tester of \fBgitcher\fR after its own author.


CHANGELOG.md
# Change log

All notable changes to this project will be documented in this file.

## v3.2 of 2019-04-18

This version adds document improvements, mailing list and homepage notices and fix some installation issues.

Note: This version has had a pair of prereleases (v3.1.2 and v3.1.3), but due to errors with the PYPI packaging they have been removed.



## v3.1.1 of 2019-10-19

This version provides the same software, but removes the repository's GitLab mirror.



## v3.1 of 2019-10-05

#### Changed

- Improve data TUI presentation.
- Correct the PGP acronym use.



## v3.0 of 2019-09-27

#### Added

- Contribution guide.


#### Changed

- Solve some bugs.
- Improve data TUI presentation.



## v2.2 of 2019-07-01

#### Added

- Test suite.


#### Changed

- Mirroring configuration. Set GitHub copy as main.



## v2.1 of 2019-06-18

#### Changed

- This version solves some performance bugs.



## v2.0 of 2019-04-28

This is the first version available to download via Homebrew.

#### Added

- Automatic setup of script endpoint to run the module directly using `$ gitcher` command.
- Interactive mode loop closure.
- Fast mode profiles listing operation.



## v1.3 of 2019-04-16

#### Added

- Profiles list representation switch.



## v1.2.2 of 2019-03-06

#### Changed

- Fix small fails with PyPI distribution.


## v1.2 of 2019-03-06

This is the first version available on PyPI.



## v1.0 of 2019-03-06

This is the first stable version of the program. It is fully tested.

#### Added

- Autocompleter helper support (only for interactive mode).


#### Changed

- Now a dictionary module recopilates every program keys.
- Escape mode: Ctrl.+C is the new method to escape. Command words like 'quit' or 'exit' will not be supported from this version.
- Relevant performance fixes and user experience improvements.



## v0.4b0 of 2019-02-23

#### Added

- Update use case option.
- Mirror use case option.
- Fast mode get current active profile option.
- Sort profiles list display.


#### Changed

- Escape command: 'q' by 'quit'. 'q' may be interesting like name or profile 
param.
- Securize and better isolate option validation.
- Relevant performance fixes.



## v0.3b0 of 2019-02-19

#### Added

- Fast mode: use shortcuts to run directly gitcher orders.
- Current git profile status.
- User input validation.
- Cherfile comments support.


#### Changed

- Securize and better isolate model layer operative.
- Fix some performance failures.




## Previous alpha phase

### v0.2a1 of 2019-02-09

#### Changed

This version fixes an error with the set up routine. The necessary libraries were not being imported.



### v0.2a0 of 2019-02-09

#### Added

- Installs the program manpages during the set up routine.


#### Changed

- Improves the software design of the project. Refactors the project as Python package and isolates GITCHER file access to a model layer.
- Improves the user experience with better output printing presentation.



### v0.1a1 of 2019-02-07

#### Changed

Set global option crashed when try to use from a directory that does not contains a git repository. It was by a performance dependencies of 'gitpython' library. The library will be removed from imports and the git commands will be played with operating system calls instead.



### v0.1a0 of 2019-02-05

#### Added

First alpha release. This version is a first approximation of the tool. It covers the entire initial scope of functionalities proposed.


gitcher.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-


"""gitcher main

gitcher is a git switcher. It facilitates the switching
between git profiles,importing configuration settings such
as name, email and user signatures.
"""

import sys
import os
from os.path import expanduser
from validate_email import validate_email

# Authorship
__author__ = "Borja González Seoane"
__copyright__ = "Copyright 2019, Borja González Seoane"
__credits__ = "Borja González Seoane"
__license__ = "GPL-3.0"
__version__ = "0.1a1"
__maintainer__ = "Borja González Seoane"
__email__ = "dev@glezseoane.com"
__status__ = "Development"

# Paths
HOME = expanduser("~")
CHERFILE = HOME + "/.cherfile"

# Prompt styles
COLOR_BLUE = '\033[94m'
COLOR_BRI_BLUE = '\033[94;1m'
COLOR_CYAN = '\033[96;1m'
COLOR_GREEN = '\033[92m'
COLOR_RED = '\033[91m'
COLOR_BOLD = '\033[1m'
COLOR_RST = '\033[0m'  # Restore default prompt style

# Predefined messages
MSG_OK = "[" + COLOR_GREEN + "OK" + COLOR_RST + "]"
MSG_ERROR = "[" + COLOR_RED + "ERROR" + COLOR_RST + "]"


# ===========================================
# =           Auxiliary functions           =
# ===========================================

# noinspection PyShadowingNames
def print_prof_error(profname):
    """function that prints a nonexistent gitcher profile error."""
    print(MSG_ERROR + " Profile {0} not exists. Try again...".format(profname))


def print_prof_list():
    """function that recuperates and prints the gitcher profile list."""
    f = open(CHERFILE, 'r')
    for line in f:
        print("-    " + COLOR_CYAN + line.split(",")[0] + COLOR_RST)


def listen(text):
    """function that listen a user input, checks if it not a 'q' (i.e.: quit
        escape command) and then canalize message to caller function. """
    reply = input(text)
    if reply == 'q':
        raise SystemExit
    else:
        return reply


def yes_or_no(question):
    """function that requires a yes or no answer"""
    reply = str(listen(question + " (y|n): ")).lower().strip()
    if reply[0] == 'y':
        return True
    if reply[0] == 'n':
        return False
    else:
        print(MSG_ERROR + " Enter (y|n) answer...")
        yes_or_no(question)


# noinspection PyShadowingNames
def check_opt(opt):
    """function that checks the integrity of the listen option."""
    return opt == 's' or opt == 'g' or opt == 'a' or opt == 'd' or opt == 'q'


# noinspection PyShadowingNames
def check_profile(profname):
    """function that checks if a gitcher profile exists."""
    f = open(CHERFILE, 'r')
    for line in f:
        if line.split(',')[0] == profname:
            return True
    return False  # if not finds prof


def check_git_context():
    """function that checks if the current directory have a git repository."""
    cwd = os.getcwd()
    return os.path.exists(cwd + "/.git")


# noinspection PyShadowingNames
def recover_prof(profname):
    """function that recovers a gitcher profile.

    Warnings:
        - check_profile must be asserted before.
        - CHERFILE can not content two profiles with the same name. The add
            function takes care of it.
    """
    f = open(CHERFILE, 'r')
    for line in f:
        words = line.split(',')
        if words[0] == profname:
            # Return as dictionary
            prof = {
                "profname": profname,
                "name": words[1],
                "email": words[2],
                "signkey": words[3],
                "signpref": words[4]
            }
            return prof


# noinspection PyShadowingNames
def switch_prof(profname, flag=''):
    """function that plays the git profile switching.

    This function can receive a '--global' flag to switch profile globally."""
    cwd = os.getcwd()  #  Current working directory path
    prof = recover_prof(profname)

    go_to_cwd = "cd {0} && ".format(cwd)
    if flag == '--global':
        go_to_cwd = ""

    cmd = "{0}git config {1} user.name '{2}'".format(go_to_cwd, flag,
                                                     prof['name'])
    os.system(cmd)

    cmd = "{0}git config {1} user.email {2}".format(go_to_cwd, flag,
                                                    prof['email'])
    os.system(cmd)

    if prof["signkey"] is not None:
        cmd = "{0}git config {1} user.signingkey {2}". \
            format(go_to_cwd, flag, prof['signkey'])
        os.system(cmd)

    # Is neccesary to run next command even preference is false because
    # 	it would be neccesary overwrite git global criteria.
    cmd = "{0}git config {1} commit.gpgsign {2}". \
        format(go_to_cwd, flag, prof['signpref'])
    os.system(cmd)


# ======================================
# =           Main launchers           =
# ======================================

# noinspection PyShadowingNames
def set_prof(profname):
    """function that sets the selected profile locally.

    It is imperative that it be called from a directory with a git
    repository."""
    if check_git_context():
        switch_prof(profname)
        print(MSG_OK + " Switched to {0} profile.".format(profname))
    else:
        print(MSG_ERROR + " Current directory not contains a git repository.")


# noinspection PyShadowingNames
def set_prof_global(profname):
    """function that sets the selected profile globally.

    It is not necessary to be called from a directory with a git repository."""
    switch_prof(profname, '--global')
    print(MSG_OK + " Set {0} as git default profile.".format(profname))


# noinspection PyShadowingNames
def add_prof():
    """function that adds a new profile."""
    print("\nLets go to add a new gitcher profile...")

    profname = listen("Enter the profile name: ")
    while check_profile(profname):
        print(MSG_ERROR + " {0} yet exists. Change name...".format(profname))
        profname = listen("Enter profile name: ")

    name = listen("Enter the git user name: ")

    email = listen("Enter the git user email: ")
    while not validate_email(email):
        print(MSG_ERROR + " Invalid email format. Try again...".format(email))
        email = listen("Enter the git user email: ")

    if yes_or_no("Do you want to use a GPG sign key?"):
        signkey = listen("Enter the git user signkey: ")
        signpref = str(yes_or_no("Do you want to autocheck every commit?"))
    else:
        signkey = None
        signpref = False

    prof = [profname, name, email, str(signkey), str(signpref)]
    prof_string = ','.join(prof)
    # Save it...
    with open(CHERFILE, 'a') as f:
        print(prof_string, file=f)


# noinspection PyShadowingNames
def delete_prof(profname):
    """function that deletes the selected profile."""
    if yes_or_no("Are you sure to delete {0}?".format(profname)):
        f = open(CHERFILE, 'r+')  # Read and write mode
        lines = f.readlines()
        lines = [line.strip('\n') for line in lines]
        f.seek(0)  # Return to the start of the file
        for line in lines:
            if line.split(',')[0] != profname:
                print(line, file=f)
        f.truncate()  # Delete possible dirty lines below
        f.close()


# ============================
# =           MAIN           =
# ============================

print(COLOR_BRI_BLUE + "**** gitcher: a git profile switcher ****" + COLOR_RST)

# First, check if CHERFILE exists and if not, propose to create it.
if not os.path.exists(CHERFILE):
    print(MSG_ERROR + " {0} not exists and it is necessary.".format(CHERFILE))
    if yes_or_no("Do you want to create {0}?".format(CHERFILE)):
        open(CHERFILE, 'w')
        print(MSG_OK + " Gitcher config dotfile created. Go on...")
    else:
        print(MSG_ERROR + " Impossible to go on without gitcher dotfile.")
        sys.exit(1)

print("gitcher profiles list:")
print_prof_list()
print("\nOptions:")
print(COLOR_CYAN + 's' + COLOR_RST + "    set a profile to current "
                                     "directory repository.")
print(COLOR_CYAN + 'g' + COLOR_RST + "    set a profile as global "
                                     "git configuration.")
print(COLOR_CYAN + 'a' + COLOR_RST + "    add a new profile.")
print(COLOR_CYAN + 'd' + COLOR_RST + "    delete a profile.")
print(COLOR_CYAN + 'q' + COLOR_RST + "    quit (escape available all time).\n")

opt = listen("Option: ")
while not check_opt(opt):
    print(MSG_ERROR + " Invalid opt! Use s|g|a|d. Type q to quit.")
    opt = listen("Enter option: ")

if not opt == 'a':
    profname = listen("Select the desired profile entering its name: ")
    while not check_profile(profname):
        print_prof_error(profname)
        profname = listen("Enter profile name: ")

    if opt == 's':
        set_prof(profname)
    elif opt == 'g':
        set_prof_global(profname)
    else:
        delete_prof(profname)
else:
    add_prof()
